File: bindings/cxx/examples/reconfigure_input_to_output.cpp
// SPDX-License-Identifier: GPL-2.0-or-later
// SPDX-FileCopyrightText: 2023 Kent Gibson <warthog618@gmail.com>

/*
 * Example of a bi-directional line requested as input and then switched
 * to output.
 */

#include <cstdlib>
#include <filesystem>
#include <gpiod.hpp>
#include <iostream>

namespace {

/* Example configuration - customize to suit your situation */
const ::std::filesystem::path chip_path("/dev/gpiochip0");
const ::gpiod::line::offset line_offset = 5;

} /* namespace */

int main()
{
	/* request the line initially as an input */
	auto request = ::gpiod::chip(chip_path)
			       .prepare_request()
			       .set_consumer("reconfigure-input-to-output")
			       .add_line_settings(
				       line_offset,
				       ::gpiod::line_settings().set_direction(
					       ::gpiod::line::direction::INPUT))
			       .do_request();

	/* read the current line value */
	::std::cout << line_offset << "="
		    << (request.get_value(line_offset) ==
					::gpiod::line::value::ACTIVE ?
				"Active" :
				"Inactive")
		    << " (input)" << ::std::endl;

	/* switch the line to an output and drive it low */
	request.reconfigure_lines(::gpiod::line_config().add_line_settings(
		line_offset,
		::gpiod::line_settings()
			.set_direction(::gpiod::line::direction::OUTPUT)
			.set_output_value(::gpiod::line::value::INACTIVE)));

	/* report the current driven value */
	::std::cout << line_offset << "="
		    << (request.get_value(line_offset) ==
					::gpiod::line::value::ACTIVE ?
				"Active" :
				"Inactive")
		    << " (output)" << ::std::endl;

	return EXIT_SUCCESS;
}

File: bindings/python/examples/reconfigure_input_to_output.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: 2023 Kent Gibson <warthog618@gmail.com>

"""Example of a bi-directional line requested as input and then switched to output."""

import gpiod

from gpiod.line import Direction, Value


def reconfigure_input_to_output(chip_path, line_offset):
    # request the line initially as an input
    with gpiod.request_lines(
        chip_path,
        consumer="reconfigure-input-to-output",
        config={line_offset: gpiod.LineSettings(direction=Direction.INPUT)},
    ) as request:
        # read the current line value
        value = request.get_value(line_offset)
        print("{}={} (input)".format(line_offset, value))
        # switch the line to an output and drive it low
        request.reconfigure_lines(
            config={
                line_offset: gpiod.LineSettings(
                    direction=Direction.OUTPUT, output_value=Value.INACTIVE
                )
            }
        )
        # report the current driven value
        value = request.get_value(line_offset)
        print("{}={} (output)".format(line_offset, value))


if __name__ == "__main__":
    try:
        reconfigure_input_to_output("/dev/gpiochip0", 5)
    except OSError as ex:
        print(ex, "\nCustomise the example configuration to suit your situation")
